File: 4UI/1ShareBlock/ProjectBlockDemo/ViewController.m
//
//  ViewController.m
//  ProjectBlockDemo
//
//  Created by lanouhn on 15/12/9.
//  Copyright (c) 2015年 lanouhn0. All rights reserved.
//

#import "ViewController.h"
#import "BlockViewController.h"


#define NSLog(format, ...) do { \
fprintf(stderr, "<%s : %d> %s\n", \
[[[NSString stringWithUTF8String:__FILE__] lastPathComponent] UTF8String], \
__LINE__, __func__); \
(NSLog)((format), ##__VA_ARGS__); \
fprintf(stderr, "-------\n"); \
} while (0)

//MyBlock --- void (^)()
typedef void (^MyBlock1)();

typedef NSString *(^Blaaa)();
//MyNewBlock1 --- int(^)(int, int)
typedef int (^MyNewBlock2)(int, int);

//MyBlock --- void (^)(NSString *)
typedef void (^MyBlock)(NSString *);

@interface ViewController ()
@property (strong, nonatomic) IBOutlet  UIView *button;
@property (nonatomic, copy) MyBlock myBlock3;//block作为属性
@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.

    //Block代码块,有个标示 ^
    //Block有两个重要元素,参数和返回值
    //1.无参无返回值:
    //声明:
    void (^myBlock1)() = ^{
        NSLog(@"A 段项目");
    };
    //调用
    myBlock1();
    
    //2.有参无返回值(Block类型中可以只写参数类型,不写参数)
    void (^myBlock2)(int, int) = ^(int value1, int value2){
        NSLog(@"%d", value1 + value2);
    };
    myBlock2(2, 3);
    
    //3.无参有返回值
    NSString *(^myBlock3)() = ^{
        return @"block";
    };
    NSString *string = myBlock3();
    NSLog(@"%@", string);
    
    //4.有参有返回值
    //返回值类型: NSString *
    //形参: (NSString *, NSString *)
    //函数名: myBlock4
    NSString *(^myBlock4)(NSString *, NSString *) = ^(NSString *str1, NSString *str2){
        return [str1 stringByAppendingString:str2];
    };
    NSString *str = myBlock4(@"I love", @" you!");
    NSLog(@"%@", str);
    
    //重定义
    MyNewBlock2 block1 = ^(int value1, int value2){
        return value1 + value2;
    };
    int sum = block1(2, 3);
    NSLog(@"%d", sum);
    
    //1.在block内部调用外部定义的变量, 是将外部变量拷贝了一份,两者的地址不同
    //2.block内部只能访问外部变量,不能修改外部变量
    //3.如果想让block内部的变量得到原本, 那么就要在前面加 __block(两个下划线)
    int a1 = 10;
    MyBlock1 myBlock5 = ^{
//        a1++;               //Block内部不能修改外部变量
        NSLog(@"a1 = %d", a1);//后调用, a1 = 10
        NSLog(@"%p", &a1);
    };
    a1 ++;
    NSLog(@"a1 == %d", a1);//先调用, a1 == 11
    NSLog(@"%p", &a1);
    //调用
    myBlock5();
    
    __block int a = 10;
    MyBlock1 myBlock6 = ^{
        a++;
        int b = a;
        NSLog(@"a = %d", b);
        NSLog(@"这是一个Block");
    };
    //调用
    myBlock6();
    
    //block循环引用  ----- 某些情况下,在block内部使用self 会产生循环引用
    //myBlock3 定义为属性会产生循环引用
//    self.myBlock3 = ^(NSString *str) {
//        self.view.backgroundColor = [UIColor redColor];
//    };
//    _myBlock3(@"haah");

    //解决循环引用   ----- 在ARC情况下,用 __weak, MRC情况下用 __block
    //弱引用(ARC下): __weak, 引用计数不加 1
    //强引用: 类似于 retain, copy, 引用计数加 1
    __weak ViewController *VC = self;
    self.myBlock3 = ^(NSString *str) {
        VC.view.backgroundColor = [UIColor redColor];
    };
    _myBlock3(@"haha");
}

//block当参数的使用
//定义一个方法,带block参数
- (void)loginWithUserID:(NSString *)userID withPassword:(NSString *)password withResult:(MyBlock)result {
    
    if ([userID isEqualToString:@"123"] && [password isEqualToString:@"123"]) {
        //登录成功
        result(@"登陆成功");//调用block
    }else {
        //登陆成功
        result(@"登陆失败,请重新输入");//调用block
    }
}

- (IBAction)handleButtonAction:(UIButton *)sender {
    
    //调用login方法
    [self loginWithUserID:@"123" withPassword:@"123" withResult:^(NSString *resultStr) {
        NSLog(@"%@", resultStr);
    }];
    
    BlockViewController *blockVc = [self.storyboard instantiateViewControllerWithIdentifier:@"Block"];
    blockVc.aBlock = ^(NSString *param1, NSString *param2) {
        self.title = [NSString stringWithFormat:@"%@%@", param1, param2];
    };
    [self.navigationController pushViewController:blockVc animated:YES];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

@end

File: 5Swift/16 Deinitialization/Deinitialization/ViewController.swift
//
//  ViewController.swift
//  Deinitialization
//
//  Created by admin on 2018/3/20.
//  Copyright © 2018年 admin. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    class Bank {
        //类型属性
        static var coinsInBank = 10_000
        //类型方法
        static func distribute(coins numberOfCoinsRequested: Int) -> Int {
            let numberOfCoinsToVend = min(numberOfCoinsRequested, coinsInBank)
            coinsInBank -= numberOfCoinsToVend
            return numberOfCoinsToVend
        }
        //类型方法
        static func receive(coins: Int) {
            coinsInBank += coins
        }
    }
    
    class Player {
        var coinsInPurse: Int
        init(coins: Int) {
            coinsInPurse = Bank.distribute(coins: coins)
        }
        func win(coins: Int) {
            coinsInPurse += Bank.distribute(coins: coins)
        }
        deinit {
            Bank.receive(coins: coinsInPurse)
        }
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
        //Deinitialization
        //关键字 deinit, 只有 Class 有 deinit, 在 dealloc 前调用
        //Swift 使用 ARC 自动管理内存
        //需要在 deinit 中手动处理的: 文件关闭, 对象销毁, KVO移除, 通知移除, NSTimer销毁
        
        //每个 Class 只能有一个 deinit, 没有参数
        //不能主动调用, 子类继承父类的 deinit 函数, 父类 deinit 函数在子类 deinit 函数调用结束后调用, 父类deinit 函数总是会被调用,即使子类的没有调用
        
        var playerOne: Player? = Player(coins: 100)
        print("A new player has joined the game with \(playerOne!.coinsInPurse) coins")
        print("There are now \(Bank.coinsInBank) coins left in the bank")
        
        playerOne!.win(coins: 2_000)
        print("PlayerOne won 2000 coins & now has \(playerOne!.coinsInPurse) coins")
        print("The bank now only has \(Bank.coinsInBank) coins left")
        
        //对象置为 nil 就会调用 Player 类中的 deinit 方法
        playerOne = nil
        print("PlayerOne has left the game")
        print("The bank now has \(Bank.coinsInBank) coins")
        
    }

    deinit {
        print("~~~~~~~~")
    }
    
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }


}

